calico_containers/pycalico/block.py

```diff
@@ -0,0 +1,383 @@
+
+    def __init__(self, cidr_prefix, host_affinity):
+        assert isinstance(cidr_prefix, IPNetwork)
+        assert cidr_prefix.cidr == cidr_prefix
+
+        # Make sure the block is the right size.
+        assert cidr_prefix.prefixlen == (BITS_BY_VERSION[cidr_prefix.version] -
+                                         BLOCK_SIZE_BITS)
+        self.cidr = cidr_prefix
+        self.db_result = None
+
+        self.host_affinity = host_affinity
+        """
+        Both to minimize collisions, where multiple hosts attempt to change a
+        single block, and to support route aggregation, each block has affinity
+        to a single Calico host.  That host does not hold exclusive rights to
+        modify the block; any host may still do that.  The host with affinity
+        simply uses the block as the place where it first searches if the user
+        asked to have the IP assigned automatically.
+        """
+
+        self.allocations = [None] * BLOCK_SIZE
+        """
+        A fixed length array with one entry for every address in the block.
+        None means unallocated.  A non-negative integer indicates the address
+        is allocated, and is the index into the `attributes` array for the
+        attributes assigned to the allocation.
+        """
+
+        self.attributes = []
+        """
+        Dictionaries of attributes for allocations.
+        """
+
+    def to_json(self):
+        """
+        Convert to a JSON representation for writing to etcd.
+        """
+
+        json_dict = {AllocationBlock.CIDR: str(self.cidr),
+                     AllocationBlock.AFFINITY:
+                         AllocationBlock.HOST_AFFINITY_T % self.host_affinity,
+                     AllocationBlock.ALLOCATIONS: self.allocations,
+                     AllocationBlock.ATTRIBUTES: self.attributes}
+        return json.dumps(json_dict)
+
+    @classmethod
+    def from_etcd_result(cls, etcd_result):
+        """
+        Convert a JSON representation into an instance of AllocationBlock.
+        """
+        json_dict = json.loads(etcd_result.value)
+        cidr_prefix = IPNetwork(json_dict[AllocationBlock.CIDR])
+
+        # Parse out the host.  For now, it's in the form host:<hostname>
+        affinity = json_dict[AllocationBlock.AFFINITY]
+        assert affinity[:5] == "host:"
+        host_affinity = affinity[5:]
+
+        block = cls(cidr_prefix, host_affinity)
+        block.db_result = etcd_result
+
+        # Process & check allocations
+        allocations = json_dict[AllocationBlock.ALLOCATIONS]
+        assert len(allocations) == BLOCK_SIZE
+        block.allocations = allocations
+
+        # Process & check attributes
+        attributes = json_dict[AllocationBlock.ATTRIBUTES]
+        block.attributes = attributes
+        assert (block._verify_attributes())
+
+        return block
+
+    def update_result(self):
+        """
+        Return the EtcdResult with any changes to the object written to
```

Comment:
Yeah, it's hard to write some abstraction layer that is both simple and very likely to work for some unknown replacement to `etcd`.  I think this code is at least simple, which should lend it to being easily unpicked should the day come that we support several backends.